Stock distribution to patient linked with invoice

mbayopanda opened the pull request:
This PR : 
- Add link between stock movement with invoice for stock exit to patient
- Add a `bhFindInvoice` component for finding invoice for a given patient
- Update the stock exit report for displaying also invoice information
- Update `End 2 End` tests for testing also the stock exit to patient with linked invoice
- Add lodash on the client side, so we can use for instance `_.groupBy()` on the client side
- Remove unused dependencies `Chart.js` and `Angular-chart`

bors[bot] committed:
Stock distribution to patient linked with invoice

link stock invoice

fix load patient from cache

load inventories in the grid

Allow to join invoice to a stock movement to a patient

test bhFindInvoice component

fix e2e tests

mbayopanda committed:
Merge branch 'master' of https://github.com/IMA-WorldHealth/bhima-2.X into link-stock-invoice

mbayopanda committed:
lint max number of rows

fix and enhancements

fix travis error

unit test for bhFindInvoice

unit-test for bhFindInvoice component

update findinvoice selector

unit-test for bhFindInvoice component

update client unit test

mbayopanda committed:
lint max number of rows

mbayopanda committed:
lint max number of rows

mbayopanda committed:
# This is a combination of 2 commits.
# This is the 1st commit message:

link invoice to stock movement

fix linting error

update voucher test

unit test for bhFindInvoice component

update unit-test for bhFindInvoice component

fix e2e test selector for findInvoice component

update unit-test for bhFindInvoice component

update client unit test

# The commit message #2 will be skipped:

# update findinvoice selector

jniles commented on client/src/js/components/bhFindInvoice.js:
```diff
@@ -0,0 +1,96 @@
+    bindings : {
+      patientUuid :       '<', // patient uuid - to restrict search to this patient
+      onSearchComplete :  '&', // bind callback to call when data is available
+      disabled :          '<', // bind disable behavior
```
This should probably be optional: `<?`

jniles commented on bower.json:
```diff
@@ -32,14 +32,13 @@
     "angular-growl-notifications": "2.6.0",
-    "webcam-directive": "3.1.1"
+    "webcam-directive": "3.1.1",
+    "lodash": "^4.17.10"
```
I don't actually see where you use `_.groupBy()`.  Where can I find it?

jniles commented on client/src/modules/templates/bhFindInvoice.tmpl.html:
```diff
@@ -0,0 +1,43 @@
+      ng-if="!$ctrl.invoiceFound && FindInvoiceForm.$submitted" 
+      class="help-block" 
+      translate="FORM.INFO.PATIENT_INVOICE_NOT_FOUND" 
+      translate-values="{{ $ctrl.translateValues }}">
```
I can't figure out where this is defined.  Did you mean to write `$ctrl.translate` to refer to [this line](https://github.com/IMA-WorldHealth/bhima-2.X/pull/2743/files#diff-e3fb58cdaea9a73c062994612f047c19R67)?

mbayopanda replied:
You're right 👍

jniles commented on client/src/js/components/bhFindInvoice.js:
```diff
@@ -0,0 +1,96 @@
+/**
+ * The Find Invoice Component
+ */
+function FindInvoiceComponent(Patients, PatientInvoice, AppCache, Notify) {
```
Do you need AppCache here?  It doesn't like you ever use it.

jniles commented on client/src/i18n/en/stock.json:
```diff
@@ -49,10 +49,13 @@
     "INCREASE_HELP"   : "Adjustment of stock by increasing the quantity of lots of products in stock",
     "INVENTORY"       : "Inventory",
     "INVENTORY_GROUP" : "Inventory Group",
+    "INVENTORIES_NOT_AVAILABLE" : "Inventories not available",
```
In English, this would be written as "Inventory not available".  The plural here is the same as the singular. 👍

jniles commented on client/src/js/app.js:
```diff
@@ -1,7 +1,7 @@
   'bhima.controllers', 'bhima.services', 'bhima.directives', 'bhima.filters',
   'bhima.components', 'bhima.routes', 'bhima.constants', 'ui.bootstrap',
-  'pascalprecht.translate', 'ngStorage', 'chart.js',
+  'pascalprecht.translate', 'ngStorage',
```
👍

mbayopanda committed:
fix test on travis

jniles commented on bower.json:
```diff
@@ -32,14 +32,13 @@
     "moment": "2.20.1",
     "angular-dynamic-locale": "0.1.32",
     "ngstorage": "0.3.11",
-    "Chart.js": "2.7.1",
```
💃

jniles commented on client/src/modules/stock/exit/exit.js:
```diff
@@ -285,6 +285,13 @@ function StockExitController(
         vm.selectableInventories = angular.copy(inventories);
+
+        // map of inventories by inventory uuid
+        vm.mapSelectableInventories = _.reduce(vm.selectableInventories, (aggregate, value) => {
```
I'm not sure it's worth adding the entire lodash library just for a _.reduce().  Isn't `vm.selectedInventories` an array?

jniles replied:
Also, you might consider making this a `new Store()`.  I think it's probably what you want - an ability to look up the entire inventory by its `uuid`.  Like this:
```js
let selectableInventoriesStore = new Store({ key : 'inventory_uuid', data : vm.selectableInventories });
```
and then later you can use it like:
```js
patient.invoice.items.forEach((item) => {
  const inventory = selectableInventoriesStore.get(item.inventory_uuid);
  if (inventory) {
    addItems(1);
    // do something
  }
});

jniles commented on client/src/modules/stock/exit/exit.js:
```diff
@@ -318,10 +325,43 @@ function StockExitController(
+    vm.stockForm.clear();
+    vm.gridApi.core.notifyDataChange(uiGridConstants.dataChange.EDIT);
+
+    if (!patient || !patient.invoice) { return; }
```
Should this `return` go before `vm.stockForm.clear()`?

jniles commented on client/src/modules/stock/exit/exit.js:
```diff
@@ -318,10 +325,43 @@ function StockExitController(
+    patient.invoice.items.forEach((item) => {
+      if (vm.mapSelectableInventories[item.inventory_uuid]) {
+        addItems(1);
+        vm.stockForm.store.data[inventoryAvailableIndex].inventory = vm.mapSelectableInventories[item.inventory_uuid];
```
Hmm..  it looks like you need to have the last item inserted into the store.  The [PatientInvoiceFormService](https://github.com/IMA-WorldHealth/bhima-2.X/blob/master/client/src/modules/invoices/patientInvoiceForm.service.js#L375) deals with this by simply returning the last element inserted into the store. I propose we do the same thing here.

Modify [the StockFormService](https://github.com/IMA-WorldHealth/bhima-2.X/blob/master/client/src/modules/stock/StockForm.service.js#L52) to `return elt;` as the last line.  Then you can write this code, which will be easier to read:
```js
patient.invoice.items.forEach((item) => {
  if (vm.mapSelectableInventories[item.inventory_uuid]) {
    const row = vm.stockForm.addItems(1);
    row.inventory = vm.mapSelecteableInventories[item.inventory_uuid];
    row.inventory_uuid = item.inventory_uuid;
    // etc..
  }
});

jniles commented on server/models/schema.sql:
```diff
@@ -1812,6 +1812,7 @@ CREATE TABLE `stock_movement` (
   `is_exit`         TINYINT(1) NOT NULL,
   `user_id`         SMALLINT(5) UNSIGNED NOT NULL,
   `reference`       INT(11) UNSIGNED NOT NULL AUTO_INCREMENT,
+  `invoice_uuid`    BINARY(16) NULL,  
```
I thought this is what the `document_uuid` was for?

mbayopanda replied:
I don't think so, `document_uuid` was for the document generated for a stock movement (SM.X.Y).

jniles replied:
The document meaning the `voucher`?

jniles commented on client/src/js/components/bhFindInvoice.js:
```diff
@@ -0,0 +1,96 @@
+    vm.invoiceFound = true;
+
+    const elementId = 'search';
+    const searchButton = document.getElementById(elementId);
```
Would it be possible to use `$document` instead of `document` here?  It will make it easier to test.

jniles commented on client/src/js/components/bhFindPatient.js:
```diff
@@ -65,16 +66,21 @@ function FindPatientComponent(Patients, AppCache, Notify, SessionService, bhCons
     });
   };
 
+  vm.$onChanges = function onChanges(changes) {
```
This change should be accompanied by a test in the [bhFindPatient](https://github.com/IMA-WorldHealth/bhima-2.X/blob/master/test/client-unit/components/bhFindPatient/controller.spec.js) controller tests.

jniles commented on client/src/modules/invoices/patientInvoice.service.js:
```diff
@@ -184,5 +190,24 @@ function PatientInvoiceService(
+  /**
+   * findInvoice
+   */
+  function findInvoice(invoiceUuid) {
```
Does this ever get used?  It doesn't look like it.

jniles commented on client/src/modules/stock/exit/exit.js:
```diff
@@ -18,8 +18,8 @@ StockExitController.$inject = [
-  Depots, Inventory, Notify, Session, util, bhConstants, ReceiptModal, StockForm, Stock,
-  StockModal, uiGridConstants, $translate, AppCache, moment, GridExportService
+  Depots, Inventory, Notify, Session, util, bhConstants, ReceiptModal, StockItem, StockForm, Stock,
+  StockModal, uiGridConstants, $translate, AppCache, moment, GridExportService, _,
```
Good use of dependency injection 👍

jniles reviewed (commented):
A few comments here.  I'll follow up with more specifics in person today.

mbayopanda committed:
fix and enhancements

mbayopanda committed:
fix and enhancements

jniles commented on client/src/modules/stock/exit/exit.js:
```diff
@@ -132,7 +132,10 @@ function StockExitController(
     data : vm.stockForm.store.data,
-    fastWatch : true,
+
+    // fastWatch to false is required for updating the grid correctly for
```
Nice comment :+1:

mbayopanda committed:
fix travis error

mbayopanda committed:
fix travis error

jniles commented on client/src/js/components/bhFindInvoice.js:
```diff
@@ -0,0 +1,96 @@
+    vm.invoiceFound = true;
+
+    const elementId = 'search';
+    const searchButton = $window.document.getElementById(elementId);
```
:+1:

jniles commented on client/src/js/components/bhFindPatient.js:
```diff
@@ -260,10 +265,10 @@ function FindPatientComponent(Patients, AppCache, Notify, SessionService, bhCons
 
     // default to findById
-    optionKey = optionKey || 'findById';
+    const _optionKey = optionKey || 'findById';
```
Just FYI, you can set default options in the function parameters.  Like this:
```js
loadDefaultOption(optionKey = 'findById') {
  findBy(optionKey);
}
```

jniles commented on client/src/modules/reports/generate/stock_exit/stock_exit.config.js:
```diff
@@ -1,12 +1,12 @@
 angular.module('bhima.controllers')
-  .controller('stock_exitController', InventoryFileConfigController);
+  .controller('stock_exitController', StockExitConfigController);
```
💃

jniles commented on gulpfile.js:
```diff
@@ -72,6 +72,10 @@ const paths = {
       'client/vendor/cropper/dist/cropper.js',
+
+      // lodash
+      'client/vendor/lodash/dist/lodash.min.js',
```
I think this should be removed.

jniles commented on server/controllers/finance/invoice/lookupConsumableInvoice.js:
```diff
@@ -0,0 +1,64 @@
+
+  let invoiceDetailQuery =
+    `SELECT
+      BUID(invoice.uuid) as uuid, CONCAT_WS('.', '${identifiers.INVOICE.key}',
```
In the case of multiple projects, this might produce the incorrect result.  I can't remember why though - just that we found it at HEV.

To retrieve the identifier, it's better to just JOIN on document_map.

jniles reviewed (commented):
Test finally pass! 👍 Well done.

It would be really good to have unit tests for `bhFindInvoice`.  They should be pretty similar to a mix between the [bhBarcodeScanner tests](https://github.com/IMA-WorldHealth/bhima-2.X/blob/master/test/client-unit/components/bhBarcoderScanner.spec.js) and the [bhFindPatient tests](https://github.com/IMA-WorldHealth/bhima-2.X/blob/master/test/client-unit/components/bhFindPatient/controller.spec.js).  Let me know if you need help with these.

mbayopanda committed:
unit test for bhFindInvoice

mbayopanda committed:
unit test for bhFindInvoice

mbayopanda committed:
unit-test for bhFindInvoice component

mbayopanda committed:
unit-test for bhFindInvoice component

mbayopanda committed:
Merge branch 'master' of https://github.com/IMA-WorldHealth/bhima-2.X into link-stock-invoice

mbayopanda committed:
update findinvoice selector

mbayopanda committed:
fix mergin conflicts

jniles commented on test/client-unit/components/bhFindInvoice.spec.js:
```diff
@@ -0,0 +1,217 @@
+    element = $compile(element)($scope);
+
+    // overwrite the getElementById function
+    _$window_.document.getElementById = (id) => find(element, `#${id}`);
```
👍

jniles commented on test/client-unit/components/bhFindInvoice.spec.js:
```diff
@@ -0,0 +1,217 @@
+  beforeEach(inject((_$rootScope_, _$compile_, _$httpBackend_, _$q_, _PatientInvoiceService_, _$window_, _NotifyService_) => {
+    $compile = _$compile_;
+    $scope = _$rootScope_.$new();
+    $httpBackend = _$httpBackend_;
```
If you never use $httpBackend for a test, you probably don't need to import it.

jniles commented on test/client-unit/components/bhFindInvoice.spec.js:
```diff
@@ -0,0 +1,217 @@
+
+  it('Runs the search method when the ENTER key is pressed', () => {
+    const ENTER_KEY = 13;
+    const FormController = $scope.$$childTail.FindInvoiceForm || {};
```
Why would this ever be empty..?

jniles commented on test/client-unit/components/bhFindInvoice.spec.js:
```diff
@@ -0,0 +1,217 @@
+    expect(angular.element(button).length).to.equal(1);
+  });
+
+  it('Runs the search method when the ENTER key is pressed', () => {
```
These are really great tests, @mbayopanda .  The descriptions are clear, each test is small and self-contained.  Really great work!

jniles commented on test/client-unit/components/bhFindInvoice.spec.js:
```diff
@@ -0,0 +1,217 @@
+    _$window_.document.getElementById = (id) => find(element, `#${id}`);
+
+    // overwrite the notification errorhandler and track it
+    Notify.handleError = () => {};
```
Why do you need to make this an empty function before making it a spy?

mbayopanda replied:
Here we want just to be notified when the `Notify.handleError` is called, and not to test the `Notify.handleError`.

jniles reviewed (approved):
Code looks pretty good.  Your unit tests are awesome!

There are 19 commits in the PR, most of which are from other PRs or merge commits or random bug fixes unrelated to the functionality here.  Could you try squashing them all into a single commit?  (If there is any confusion, [this SO answer](https://stackoverflow.com/a/5189600) talks about squashing).  That way we have a more focused git history when we eventually do our next release.

mbayopanda committed:
removing not necessary dependencies

mbayopanda committed:
fix linting error

bors[bot] committed:
link invoice to stock movement

fix linting error

update voucher test

unit test for bhFindInvoice component

update unit-test for bhFindInvoice component

fix e2e test selector for findInvoice component

update unit-test for bhFindInvoice component

update client unit test

mbayopanda committed:
remove overwrite handleError on bhFindInvoice unit test

mbayopanda committed:
fix mergin conflicts

mbayopanda committed:
Merge branch 'save2-link-stock-invoice' into link-stock-invoice

mbayopanda committed:
set mocha bail to false

mbayopanda committed:
update voucher e2e test
